File: cpp/animal.cpp
#include <iostream>
#include <string>
#include <unordered_map>

std::unordered_map<long long, std::string> data = {{1, "Оно умеет плавать?"}, {2, "птица"}, {3, "рыба"}};

bool yesNo(const std::string &question)
{
    while (true)
    {
        std::cout << question << " (д/н) ";
        std::string s;
        std::getline(std::cin, s);
        if (s == "д" || s == "Д")
        {
            return true;
        }
        if (s == "н" || s == "Н")
        {
            return false;
        }
    }
}

int main()
{
    do
    {
        std::cout << "\nЗагадайте животное...\n";
        long long i = 1;
        while (true)
        {
            std::string current = data[i];
            long long no = i * 2;
            long long yes = i * 2 + 1;
            if (current.back() == '?')
            {
                if (yesNo(current))
                {
                    i = yes;
                }
                else
                {
                    i = no;
                }
            }
            else
            {
                if (yesNo("Это " + current + "?"))
                {
                    std::cout << "Ура, я отгадал!\n";
                }
                else
                {
                    std::cout << "Вы победили!\nЧто за животное вы загадали? ";
                    std::string animal;
                    std::getline(std::cin, animal);
                    std::cout << "С помощью какого вопроса можно отличить "
                        << current << " от " << animal << "? ";
                    std::string question;
                    std::getline(std::cin, question);
                    data[i] = question;
                    data[no] = current;
                    data[yes] = animal;
                }
                break;
            }
        }
    }
    while (yesNo("Хотите сыграть ещё?"));
}

File: python/animal.py
#!/usr/bin/python3

data = {1: 'Оно умеет плавать?', 2: 'птица', 3: 'рыба'}

def yesNo(question):
    while True:
        s = input(f'{question} (д/н) ')
        if s == 'д' or s == 'Д':
            return True
        if s == 'н' or s == 'Н':
            return False

while True:
    print('\nЗагадайте животное...')
    i = 1
    while True:
        current = data[i]
        no = i * 2
        yes = i * 2 + 1
        if current[-1] == '?':
            if yesNo(current):
                i = yes
            else:
                i = no
        else:
            if yesNo(f'Это {current}?'):
                print('Ура, я отгадал!')
            else:
                animal = input('Вы победили!\nЧто за животное вы загадали? ')
                question = input(f'С помощью какого вопроса можно отличить {current} от {animal}? ')
                data[i] = question
                data[no] = current
                data[yes] = animal
            break
    if not yesNo('Хотите сыграть ещё?'):
        break
